### ISSUE TITLE
Asynchronism in the generation of images

### ISSUE THREAD

**Issue by rdhb17** [bpampuch/pdfmake#1361]
Hello, good evening, I am generating a report with PDFmake and this report contains images that I generate with Flot, among others. The fact is that the report is generated correctly in Windows XP using Chrome. However, with Firefox in Windows XP there are delays in the generation of the pages of the report and insert images in disorder. Something similar happens with Chrome and Firefox in Windows 7 and Windows 10, where even many of these images are not even included in the report. We know that JavaScript is an asynchronous language and I am using callback instructions for the commands to be executed sequentially.

Please, if you can give me some ideas to solve this asynchronism issue to correctly insert the images within the report.

**Comment by liborm85** [comment]
Attach some example for reproduce problem.

**Comment by rdhb17** [comment]
Thank you very much for the quick answer. The code of this program is quite large. So I'm doing a summary of the functions committed to these delays. Below you can see the code of this summary.

/* 
Generation of a report using PDFmake and Flot

This code runs perfectly on Windows XP using Chrome. However, with Firefox in Windows XP delays 
in the generation of images begin to occur. There are also delays in Windows 7 and Windows 10 in both 
Chrome and Firefox.
*/

// Here goes the definition of many global variables.

// Through this function we start generating the report.
function Synchro() {
	var startDate = new Date();
	var startTime = startDate.getTime();
	Synchronize(function(dd) {
		pdfMake.createPdf(dd).open();
	});
	var endDate = new Date();
	var endTime = endDate.getTime();
	var elapsed = (endTime - startTime);
	$("#tlapso").text('Time elapsed: '+elapsed+'ms');
	return;
}

function Synchronize(callback) {
//  Here are variable definitions and code for report headers which is generated correctly
//  ... more code here...

	// This function is what seems to take time to generate the image.
	// This is the part where the first page is generated where the images are included after the headings
	ComposeGraph(nFrom,nFrom,nTo,actDim1,0,function(imgURL) {
		imgPAG1 = imgURL;
	});

	columns = {
		columns: [ 
			{
				image: imgPAG1, // here the generated image is inserted
				width: 414,
				height: actDim2,
				relativePosition: {x: 0, y: 0}
			},
			{
				width: '*',
				margin: [3,0,0,0],
				stack: stk
			}
		],
		absolutePosition: {x: 24, y: nextPosY},
		columnGap: 0
	};
	cnt.push(columns);

	//  ... more code here...

	// This is where the other pages where images are inserted in the report are generated.

	//  ... more code here...

		for (i = 1; i <= totalPags; i++) {
			//  ... more code here...

			// This function is what seems to take time to generate the images.
			ComposeGraph(nFrom,nFrom,nTo,1584,0,function(imgURL) {
				imgPAG2 = imgURL;
			});
			columns = {
				columns: [ 
					{
						image: imgPAG2, // here the generated image is inserted
						width: 414,
						height: 792,
						relativePosition: {x: 0, y: 0}
					},
					{
						width: '*',
						margin: [3,0,0,0],
						stack: stk
					}
				],
				absolutePosition: {x: 24, y: 0},
				columnGap: 0
			};
			cnt.push(columns);
			
		}
		
		//  ... more code here...

	// Here the array of the content of the report is generated.
	var dd = { 
		pageSize: 'LETTER', pageOrientation: 'portrait', pageMargins: [ 20, 0, 24, 0 ],
		content: cnt,
		styles: {
			maintitle: { fontSize: 16, color: 'black', bold: true, alignment: 'center' },
			footer: { fontSize: 10, alignment: 'center' },
			comments: { italics: true, color: '#F00' }
		},
		defaultStyle: {
			color: 'gray',
			columnGap: 0,
		}
	};

	// Here we return from the Synchronize function by executing the callback function
	callback(dd);
}

function ComposeGraph(mt,nFrom,nTo,ydim,yoff,callback) {
	// I am using these image objects to use the onload event with better result than simply using the new image object
	$("<img onload='ImgLoaded()' src='' id='tmpImg0'>").css({
		position: "absolute",
		display: "none",
		top: "0px",
		left: "0px",
	}).prependTo("body");
	// ... here more image objects ...
	$("<img onload='ImgLoaded()' src='' id='tmpImg6'>").css({
		position: "absolute",
		display: "none",
		top: "0px",
		left: "0px",
	}).prependTo("body");

	NumImg = 0; // This variable is global.
	// here we load the images previously drawn with Flot
	tmpImg0.src = flot0.getCanvas().toDataURL("image/png");
	// ... Here we load more images...

	// This function Wait until all the images have loaded.
	var Loading = setInterval(function () {
		if (NumImg >= 6) {
			clearInterval(Loading);
		}
	},50000);
	// This function creates a canvas and makes some drawings and returns the canvas
	var canvas = ComposeGraph0(mt,nFrom,nTo,ydim,yoff);
	var ctx    = canvas.getContext('2d');
	// Here we insert the images generated with Flot
	ctx.drawImage(tmpImg0, 0, 0);
	// ... Here we draw more images...

	NumImg = 0; // This variable is global.
	// here we load the complete canvas with all the images and drawings included
	tmpImg6.src = canvas.toDataURL("image/png");
	// here we wait for the complete image to be loaded
	Loading = setInterval(function () {
		if (NumImg >= 1) {
			clearInterval(Loading);
		}
	},50000);

	// Here we return the complete image.
	callback(tmpImg6.src);
}

// This function is executed when the onload event is called.
function ImgLoaded() {
	NumImg++;
	return;
}

**Comment by liborm85** [comment]
I see there `setInterval` which can cause delays. But code is very confusing and isn't runnable. Create simple example runnable on [playground](http://pdfmake.org/playground.html) or [jsfiddle.net](https://jsfiddle.net/). Without simulation I have no way to explore it.

**Comment by rdhb17** [comment]
It is not possible to create an executable example because this program interacts with databases through Ajax.

The setInterval instruction has the purpose of waiting for the onload event to complete loading all the images before creating each page. But apparently setInterval does not work and instead shows empty or incomplete images.

Remember that this code runs correctly in Windows XP using Chrome.

**Comment by liborm85** [comment]
The problem is in this, see below:
```js
// ...
		for (i = 1; i <= totalPags; i++) {
			//  ... more code here...

			// This function is what seems to take time to generate the images.
			ComposeGraph(nFrom,nFrom,nTo,1584,0,function(imgURL) {
				imgPAG2 = imgURL;
			});
```
**Here isn't available image in `imgPAG2` variable, because `ComposeGraph` is asynchronous. Therefore, no image is inserted in the following code.**
```js
			columns = {
				columns: [ 
					{
						image: imgPAG2, // here the generated image is inserted
						width: 414,
						height: 792,
						relativePosition: {x: 0, y: 0}
					},
					{
						width: '*',
						margin: [3,0,0,0],
						stack: stk
					}
				],
				absolutePosition: {x: 24, y: 0},
				columnGap: 0
			};
			cnt.push(columns);
			
		}
// ...
```

You need to make asynchronous whole `docDefinition` generation.
This issue is not related to pdfmake, but general javascript.

**Comment by rdhb17** [comment]
After dealing with this problem for several days I was finally able to solve it. It's true it has nothing to do with PDFmake, it's more of a javascript problem.

Now I want to comment on the solution in this medium because maybe it will help someone. The problem was in the drawImage statement and the solution was to simply use another equivalent instruction putImageData.

For this I wrote the following function:

function drawImage (plot, x, y, w, h) {
     var ctx = plot.getCanvas (). getContext ('2d');
     var imgData = ctx.getImageData (0, 0, w, h);
     ctxGraph.putImageData (imgData, x, y); // ctxGraph global context variable
     return;
}

I made some other changes in the corresponding functions and everything started to work correctly.

Thank you very much for having attended my request for collaboration. Best regards to all.
